File: training-z/src/app/modules/workout-planner/models/training-plan.ts
import { TrainingUnit } from './training-unit';

export class TrainingPlan {
  id: string;
  name: string;
  trainingUnits: TrainingUnit[];
  isActive: boolean;

  constructor(
    id: string,
    name: string,
    trainingUnits: TrainingUnit[],
    isActive: boolean
  ) {
    this.id = id;
    this.name = name;
    this.trainingUnits = trainingUnits;
    this.isActive = isActive;
  }

  static fromBackend(raw: any): TrainingPlan {
    return new TrainingPlan(
      raw.id,
      raw.name,
      raw.trainingUnits.map((u: any) => TrainingUnit.fromBackend(u)),
      raw.isActive
    );
  }
}

File: TrainingZ.Domain/Entities/TrainingPlan.cs
using TrainingZ.Domain.Abstract;

namespace TrainingZ.Domain.Entities;

public class TrainingPlan : BaseEntity
{
    public Guid CoachingDataId { get; set; }
    public CoachingData? CoachingData { get; set; }
    public DateTime LastModified { get; set; }
    public string Name { get; set; } = string.Empty;
    public bool IsActive { get; set; }
    public ICollection<TrainingUnit> TrainingUnits { get; set; } = [];

    public TrainingPlan() {}

    public TrainingPlan(Guid coachingDataId, DateTime lastModified, string name, bool isActive, DateTime createdAt)
    {
        CoachingDataId = coachingDataId;
        LastModified = lastModified;
        Name = name;
        CreatedAt = createdAt;
        IsActive = isActive;
    }

    public TrainingPlan DeepCopyWithoutInclusions()
    {
        return new TrainingPlan()
        {
            Id = Id,
            CreatedAt = CreatedAt,
            LastModified = LastModified,
            Name = Name,
            IsActive = IsActive,
            TrainingUnits = TrainingUnits.Select(x => x.DeepCopyWithoutInclusions()).ToList()
        };
    }
}
